### ISSUE TITLE
Make KCP more resilient to flaky networking

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/cluster-api#6657]
KCP by design prefers stability over speed, and this in practice is translated into KCP requiring a set of information to be present before taking any action no matter the operation he is currently performing (scaling up, down, upgrade, remediation, etc.).

That information is collected from the workload cluster, and in order to do so a KCP depends on a fully working network setup.

Based on observations on several production clusters, it appears that the most complex/error-prone connection KCP requires is the connection from the KCP to the etcd instances on control plane nodes. Two factors make this connection tricky:
- The connection relies on api-server proxy capabilities to get access on etcd running on a machine CAPI doesn't get access to, and this stress several layers of networking (KCP manager -> workload cluster's api-server, workload cluster's api-server -> kubelet, kubelet->pod)
- This connection is probed in several ways from KCP, in order to check the state of each etcd member and detect eventual split-brain conditions.

This issue is about exploring some ideas to allow KCP to:
- be more resilient to temporary problems when connecting to etcd
- recovery faster from temporary etcd connection unavailability

NOTE-1: This work is not intended to remove the need for a working network setup, but only to improve the overall system resiliency to temporary problems. 
NOTE-2: This work is not intended to change the current network setup required by CAPI (we are not exploring different ways for KCP (CAPI) to collect info from the workload cluster).

/kind feature
/area control-plane

/assign

**Comment by josh-ferrell** [comment]
Linking #6765

**Comment by josh-ferrell** [comment]
@fabriziopandini One idea I had was to add logic that would have the KCP manager -> api-server connection use the internal kubernetes.default.svc connection in self-hosted clusters. In theory this would circumvent any management cluster LB issues regardless of provider.

**Comment by sbueringer** [comment]
Probably related: https://github.com/kubernetes-sigs/cluster-api/issues/6337

**Comment by fabriziopandini** [comment]
> This issue is about exploring some ideas to allow KCP to:
>
> - be more resilient to temporary problems when connecting to etcd
> - recovery faster from temporary etcd connection unavailability

I investigated two options:
- adding retry on etcd connection creation, but this did not provide improvements on a system with flaky networking, it just increased the time to fail.
- forcing KCP to reconcile after detecting some etcd connection problem, but this only increased the rate of flakiness being reported.

Now, we have identified the root cause of flakiness, which was related to api server pinning on CP nodes for CAPZ - which works at machine level but not at pod level / when going through Core DNS

> One idea I had was to add logic that would have the KCP manager -> api-server connection use the internal kubernetes.default.svc connection in self-hosted clusters. In theory this would circumvent any management cluster LB issues regardless of provider.

using different load balancers for self-managed clusters/other use cases, is an interesting idea but IMO it requires some design/research work. top of mind

- CAPI does not have a well-defined notion - something we can use in the code - of “this is is a management cluster” or “this is a self-hosted cluster”, so this should be introduced from scratch. 
- There are different issues about kubeconfig to be used by CAPI or by users, like the one linked by @sbueringer, and we really need to consider them together in order to avoid a fragmented solution
